### PR TITLE
[SPARK-42643][CONNECT][PYTHON] Register Java (aggregate) user-defined functions

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Implement `spark.udf.registerJavaFunction` and `spark.udf.registerJavaUDAF`.
 A new proto `JavaUDF` is introduced.

### Why are the changes needed?
Parity with vanilla PySpark.

### Does this PR introduce _any_ user-facing change?
Yes. `spark.udf.registerJavaFunction` and `spark.udf.registerJavaUDAF` are supported now.

### How was this patch tested?
Parity unit tests.

SPARK-41661

### CODE REVIEW

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 1]
```diff
@@ -303,14 +303,15 @@ message Expression {
 message CommonInlineUserDefinedFunction {
   // (Required) Name of the user-defined function.
   string function_name = 1;
-  // (Required) Indicate if the user-defined function is deterministic.
+  // (Optional) Indicate if the user-defined function is deterministic.
   bool deterministic = 2;
```
Comment (xinrong-meng) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:307]: JavaUDF has no `deterministic` field but the server doesn't have logic that relies on that field. So only the comment is changed.

**Comment by xinrong-meng** [comment]
CC @zhengruifeng @HyukjinKwon @ueshin @hvanhovell

**Comment by xinrong-meng** [comment]
After double thoughts, I add `spark.udf.registerJavaUDAF`'s implementation into this PR, since both APIs rely on the same proto message JavaUDF.

**Comment by HyukjinKwon** [comment]
WDYT @hvanhovell ?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by xinrong-meng** [comment]
Merged to master and branch-3.4, thanks all!

**Pull request closed without merging**